### PR TITLE
Use std::map as a return value for Response::headers()

### PR DESCRIPTION
**Pull request by FolMing**
Please use [RFC 7230, section 3.2.2](https://www.rfc-editor.org/rfc/rfc7230.html#section-3.2.2) to fully understand the changes made in this PR!

### CODE REVIEW

File: eventuals/builder.h [hunk 1]
```diff
@@ -117,6 +123,12 @@ class FieldWithDefault<Value_, true> : public Field<Value_, true> {
     : Field<Value_, true>(std::forward<Value>(value)) {}
 
   FieldWithDefault(FieldWithDefault&&) = default;
+
+  template <typename Value>
+  auto Set(Value&& value) {
+    static_assert(std::is_convertible_v<Value, Value_>);
+    return FieldWithDefault<Value_, true>{std::forward<Value>(value)};
+  }
```
Comment (FolMing) [inline eventuals/builder.h:131]: Apparently, we could not use multiple Request::_Builder::header() before, so I had to do this change. I did the same for `Field<Value_, true>` as well for consistency.

File: eventuals/http.h [hunk 2]
```diff
@@ -547,9 +566,91 @@ struct _HTTP {
...
+                          // Cookie: cookie1=value1, cookie2=value2
+                          auto iterator = headers.find(key);
+                          if (iterator == headers.end()) {
+                            // Header doesn't exist yet.
+                            headers.emplace(std::move(key), std::move(value));
+                          } else {
+                            // Header already exists.
+                            headers[key] += ", ";
+                            headers[key] += std::move(value);
+                          }
```
Comment (FolMing) [inline eventuals/http.h:648]: This one is large, but should be self-explanatory.

File: test/http.cc [hunk 3]
```diff
@@ -52,7 +52,10 @@ TEST_P(HttpTest, Get) {
   auto response = future.get();
 
   EXPECT_EQ(200, response.code());
-  EXPECT_THAT(response.headers(), testing::ContainsRegex("Foo: Bar"));
+  EXPECT_THAT(
+      response.headers(),
+      testing::Contains(http::Header("Foo", "Bar")));
+  EXPECT_EQ(response.headers().at("Content-Length"), "25");
```
Comment (FolMing) [inline test/http.cc:58]: Even if we are using std::map, we can't use operator[] as response.headers() gives back a constant reference!

File: eventuals/http.h [hunk 4]
```diff
@@ -180,7 +181,25 @@ class Request::_Builder : public builder::Builder {
...
+    // A sender MUST NOT generate multiple header fields with the same field
+    // name in a message unless either the entire field value for that
+    // header field is defined as a comma-separated list [i.e., #(values)]
+    // or the header field is a well-known exception (as noted below).
+    //
+    // NOTE: If user tries to add an already existing header,
+    // append the new one to the old one using comma.
+    // Example:
+    // Cookie: cookie1=value1, cookie2=value2
+    auto iterator = headers_.value().find(key);
```
Comment (FolMing) [inline eventuals/http.h:194]: Would have used std::map::contains(), but it's a C++20 method.

**Review by FolMing — COMMENTED** [review]
A comment to help you guys understand the changes!

File: eventuals/http.h [hunk 5]
```diff
@@ -547,9 +566,91 @@ struct _HTTP {
...
+                        // HTTP/1.1 200
+                        // SomeHeaderKey1: SomeHeaderValue1
+                        // SomeHeaderKey2: SomeHeaderValue2
+                        // --------------------------------
+                        while (!headers_buffer_stringstream.eof()) {
+                          std::string line;
+                          std::getline(headers_buffer_stringstream, line);
+
+                          // Find where ':' is.
+                          auto column_pos = std::find(
```
Comment (benh) [inline eventuals/http.h:586]: What's a `pos`? 🤪 Keep me honest, I'm trying to be use complete words too but sometimes I forget!

File: eventuals/http.h [hunk 6]
```diff
@@ -547,9 +566,91 @@ struct _HTTP {
...
+                          if (column_pos == line.cend()) {
+                            continue;
+                          }
+
+                          // Assign key and value.
+                          auto key = std::string(line.cbegin(), column_pos);
+                          auto value = std::string(column_pos + 1, line.cend());
+
+                          // Helper lambda for removing leading
+                          // and trailing spaces.
```
Comment (benh) [inline eventuals/http.h:601]: Let's add a `TODO` here to bring in the [`absl`](https://abseil.io) library which does this all for us. We bring it in for `stout-flags` [here](https://github.com/3rdparty/stout-flags/blob/main/bazel/deps.bzl#L14).

File: test/http.cc [hunk 7]
```diff
@@ -231,6 +234,56 @@ TEST_P(HttpTest, GetHeaders) {
...
```
Comment (benh) [inline test/http.cc:240]: Love seeing usage of `testing::Contains()` above, let's also do that here!

File: test/http.cc [hunk 8]
```diff
@@ -231,6 +234,56 @@ TEST_P(HttpTest, GetHeaders) {
...
+
+  EventLoop::Default().RunUntil(future);
+
+  auto response = future.get();
+
+  EXPECT_EQ(200, response.code());
+  EXPECT_THAT(
+      response.headers(),
+      testing::Contains(http::Header("Foo", "Bar1, Bar2")));
+  EXPECT_EQ(response.headers().at("Content-Length"), "25");
```
Comment (benh) [inline test/http.cc:287]: Ditto using `testing::Contains()` here.

File: test/http.cc (hunk 3 above)
Comment (benh) [reply test/http.cc:58]: Makes sense, but why not use `testing::Contains()` like you do on the previous line? It's safer and the preferred method of doing it.

File: test/http.cc [hunk 9]
```diff
@@ -231,6 +234,56 @@ TEST_P(HttpTest, GetHeaders) {
...
+  http::Client client = server.Client();
+
+  EXPECT_CALL(server, ReceivedHeaders)
+      .WillOnce([](auto socket, const std::string& data) {
+        EXPECT_THAT(data, testing::ContainsRegex("foo: bar1, bar2"));
+
+        socket->Send(
+            "HTTP/1.1 200 OK\r\n"
+            "Foo: Bar1\r\n"
+            "Foo: Bar2\r\n"
```
Comment (benh) [inline test/http.cc:260]: Based on the RFC that you commented on above, doesn't this violate the rule? Clearly curl handles it given this test passes, but is that what you're trying to test? Or were you just trying to test multiple header values?

**Review by benh — CHANGES_REQUESTED** [review]
Most of this looks great! Let me know if you'd like to discuss the `Repeated` comments synchronously, otherwise for now I'm happy to just have a builder field of type `Headers` instead of wrapping it in a `builder::FieldWithDefault`.

File: test/http.cc (hunk 3 above)
Comment (FolMing) [reply test/http.cc:58]: @benh it's a showcase for other people, who will use our library.

File: test/http.cc (hunk 7 above)
Comment (FolMing) [reply test/http.cc:240]: Again, just a showcase for others!

File: test/http.cc (hunk 8 above)
Comment (FolMing) [reply test/http.cc:287]: ^

File: test/http.cc (hunk 9 above)
Comment (FolMing) [reply test/http.cc:260]: From what I read I understand that server "can" do this and we, as a client, "must not" do this.
Take a look: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie

File: eventuals/builder.h [hunk 10]
```diff
@@ -29,6 +29,12 @@ class Field<Value_, true> {
   Field(Value_ value)
     : value_(std::move(value)) {}
 
+  template <typename Value>
+  auto Set(Value&& value) {
+    static_assert(std::is_convertible_v<Value, Value_>);
+    return Field<Value_, true>{std::forward<Value>(value)};
+  }
+
```
[the remark this answers is not in the document]
Comment (FolMing) [reply eventuals/builder.h:37]: `ComboField`? :D

File: eventuals/builder.h (hunk 1 above)
Comment (FolMing) [reply eventuals/builder.h:131]: I personally think going with just `Headers` type is the way to handle such cases, dunno if we really want to wrap things up.

**Review by rjhuijsman — COMMENTED** [review]
I like the API of having headers be an `std::map` a lot! 

I don't know this code well enough to give more than that surface-level review though, so I'll leave the real reviewing to Ben. 🙂

**Review by benh — APPROVED** [review]
Thanks Nik!

**Pull request merged**